Software/DumptruckUltra-MCU/logic/driving-algorithm/DrivingAlgorithm.cpp

```diff
@@ -0,0 +1,130 @@
+    : leftMotor{driveMotors.leftMotor},
+      rightMotor{driveMotors.rightMotor},
+      getFrontDistanceFunction{std::move(getFrontDistanceFunction)},
+      getPoseFunction{std::move(getPoseFunction)},
+      currentTarget{.x = 0, .y = 0, .heading = 0},
+      currentStatus{DrivingAlgorithmStatus::STOPPED},
+      drivingTaskHandle{} {
+
+    // Need valid functions
+    CY_ASSERT(this->getFrontDistanceFunction != nullptr);
+    CY_ASSERT(this->getPoseFunction != nullptr);
+
+    // Create FreeRTOS task
+    const auto drivingTaskSetupResult{
+        xTaskCreate(
+            [](void *obj) { static_cast<DrivingAlgorithm *>(obj)->drivingTask(); },
+            DRIVING_ALGORITHM_TASK_NAME,
+            DRIVING_ALGORITHM_STACK_SIZE,
+            this,
+            DRIVING_ALGORITHM_PRIORITY,
+            &drivingTaskHandle)};
+    CY_ASSERT(drivingTaskSetupResult == pdPASS);
+}
+
+auto DrivingAlgorithm::getStatus() const -> DrivingAlgorithmStatus {
+    return currentStatus;
+}
+
+void DrivingAlgorithm::loadNewTarget(const DeadReckoning::Pose2D &position) {
+    currentTarget = position;
+}
+
+void DrivingAlgorithm::start() {
+    leftMotor.enable();
+    rightMotor.enable();
+    currentStatus = DrivingAlgorithmStatus::RUNNING;
+    // Manipulate task LAST to avoid scheduling shennanigans
+    vTaskResume(drivingTaskHandle);
+}
+
+void DrivingAlgorithm::stop(const DrivingAlgorithmStatus &stopStatus) {
+    leftMotor.disable();
+    rightMotor.disable();
+    currentStatus = stopStatus;
+    // Manipulate task LAST to avoid scheduling shennanigans
+    vTaskSuspend(drivingTaskHandle);
+}
+
+void DrivingAlgorithm::stop() {
+    stop(DrivingAlgorithmStatus::STOPPED);
+}
+
+void DrivingAlgorithm::drivingTask() {
+
+    // No setup
+
+    while (true) {
+        const auto currPose{getPoseFunction()};
+        // If we are not at the target...
+        if (distanceToTarget(currPose) > TARGET_PROXIMITY_THRESHOLD_METERS || (currentTarget.heading - currPose.heading) > TARGET_HEADING_THRESHOLD_RADIANS) {
+            // Find target heading (if we are already close to the target, turn to match the goal heading)
+            const float headingToTarget{(distanceToTarget(currPose) > TARGET_PROXIMITY_THRESHOLD_METERS)
+                                            ? std::atan2(currentTarget.y - currPose.y, currentTarget.x - currPose.x)
+                                            : currentTarget.heading};
+            // If something is in the way...
+            if (getFrontDistanceFunction() < DISTANCE_THRESHOLD_METERS) {
+                // Turn right
+                leftMotor.setPower(Hardware::Motors::Motor::MOTOR_MAX_SPEED_ABS);
+                rightMotor.setPower(-Hardware::Motors::Motor::MOTOR_MAX_SPEED_ABS);
+            } else {
+                // Otherwise, drive at the target
+                const auto motorPowers{deltaAngleToDrivePowers(headingToTarget - currPose.heading)};
```

Comment:
currPose.heading is updated on Line 73, and headingToTarget may be equal to it before it starts turning. Then, the deltaAngleToDrivePowers would be 0. I believe you have to call getPoseFunction() again before computing deltaAngle.

Reply:
That's fine.  The `deltaAngleToDrivePowers()` function will return motor powers that drive forward in that case.  Since the `headingToTarget` and `currPose.heading` are aligned (since their difference is 0), then we want to drive straight forward anyways.